### PR TITLE
SOLR-15795: Fix REPLACENODE to not use source node

### PR DESCRIPTION
**Pull request by HoustonPutman**
Add test to make sure it fails correctly.

https://issues.apache.org/jira/browse/SOLR-15795

I also want to make this better, so that if you are moving multiple replicas, the placement formulas take into account where the previous replicas were placed when deciding the next replica placement. Maybe this deserves its own Jira though.

### CODE REVIEW

**Comment by HoustonPutman** [comment]
@sigram , you might be interested too, but I can't add you as a reviewer.

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java [hunk 1]
```diff
@@ -356,6 +356,26 @@ public int weight() {
...
+    boolean anyLiveNodes = false;
+    for (String node : createNodeList) {
+      anyLiveNodes |= liveNodes.contains(node);
+    }
+    if (!anyLiveNodes) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+          "None of the node(s) specified " + createNodeList + " are not currently active in "
+              + createNodeList + ", no action taken.");
+    }
+    return createNodeList; // unmodified, but return for inline use. Only modified if empty, and that will throw an error
```
Comment (madrob) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:376]: is this meaningful? I don't see us using it yet in the patch.

**Review by madrob — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java (hunk 1 above)
Comment (HoustonPutman) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:376]: It's used down below to replace the assert.

File: solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java [hunk 2]
```diff
@@ -210,6 +212,25 @@ public void test() throws Exception {
...
+    if (log.isInfoEnabled()) {
+      log.info("total_jettys: {}", cluster.getJettySolrRunners().size());
+    }
+
+    CloudSolrClient cloudClient = cluster.getSolrClient();
+    cloudClient.request(CollectionAdminRequest.createCollection(coll, "conf1", 5, 1,0,0));
+
+    cluster.waitForActiveCollection(coll, 5, 5);
+
+    String liveNode = cloudClient.getZkStateReader().getClusterState().getLiveNodes().stream().findAny().get();
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:230]: Just remembered the reason I use this. `getLiveNodes()` returns a set, which has no equivalent to `get(0)`

[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:230]: Nvm, just read the second half of your comment. Will fix.

File: solr/core/src/java/org/apache/solr/cloud/api/collections/ReplaceNodeCmd.java [hunk 3]
```diff
@@ -115,7 +119,7 @@ public void call(ClusterState state, ZkNodeProps message, NamedList<Object> resu
               .assignNrtReplicas(numNrtReplicas)
               .assignTlogReplicas(numTlogReplicas)
               .assignPullReplicas(numPullReplicas)
-              .onNodes(new ArrayList<>(ccc.getSolrCloudManager().getClusterStateProvider().getLiveNodes()))
+              .onNodes(ccc.getSolrCloudManager().getClusterStateProvider().getLiveNodes().stream().filter(node -> !node.equals(source)).collect(Collectors.toList()))
```
Comment (murblanc) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/ReplaceNodeCmd.java:122]: Can you please remove line 185 and the empty finally below while you're changing this file (unrelated to your changes and Github doesn't allow commenting there)

File: solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java [hunk 4]
```diff
@@ -158,12 +165,7 @@ public void test() throws Exception {
     // make sure all newly created replicas on node are active
     List<Replica> newReplicas = collection.getReplicas(node2bdecommissioned);
     replicas.forEach(r -> {
```
Comment (murblanc) [inline solr/core/src/test/org/apache/solr/cloud/ReplaceNodeTest.java:167]: Curly brackets are not needed:

`replicas.forEach(r -> newReplicas.removeIf(nr -> nr.getName().equals(r.getName())));`

**Review by murblanc — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java [hunk 5]
```diff
@@ -356,6 +356,26 @@ public int weight() {
...
+    if (createNodeList == null) {
+      createNodeList = Collections.emptyList();
+    }
+    boolean anyLiveNodes = false;
+    for (String node : createNodeList) {
+      anyLiveNodes |= liveNodes.contains(node);
+    }
+    if (!anyLiveNodes) {
+      throw new SolrException(SolrException.ErrorCode.BAD_REQUEST,
+          "None of the node(s) specified " + createNodeList + " are not currently active in "
```
[the remark this answers is not in the document]
Comment (HoustonPutman) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/Assign.java:373]: Good find!

File: solr/core/src/java/org/apache/solr/cloud/api/collections/ReplaceNodeCmd.java (hunk 3 above)
Comment (HoustonPutman) [reply solr/core/src/java/org/apache/solr/cloud/api/collections/ReplaceNodeCmd.java:122]: have to remove the whole try{} as well, will do.

**Pull request merged**